server/resp/src/main/java/org/infinispan/server/resp/RespLettuceHandler.java

```diff
@@ -32,25 +39,146 @@ public RespLettuceHandler(RespRequestHandler initialHandler) {
+            type = StringCodec.ASCII.decodeKey(bytes);
+            return;
+         }
+
+         if (!initialized) {
+            output = new ArrayList<>();
+         }
+
+         output.add(bytes == null ? null : codec.decodeValue(bytes));
+      }
+
+      @Override
+      public void setSingle(ByteBuffer bytes) {
+         if (type == null) {
+            set(bytes);
+         }
+         if (!initialized) {
+            output = new ArrayList<>();
+         }
+
+         output.add(bytes == null ? null : StringCodec.UTF8.decodeValue(bytes));
+      }
+
+      public String type() {
+         return type;
+      }
+
+      public List<Object> getContent() {
+
+         List<Object> copy = new ArrayList<>();
+
+         for (Object o : get()) {
+            if (o instanceof ByteBuffer) {
+               copy.add(((ByteBuffer) o).asReadOnlyBuffer());
+            } else {
+               copy.add(o);
+            }
+         }
+
+         return Collections.unmodifiableList(copy);
+      }
+
+      // Copied from NestedMultiOutput
+
+      private final Deque<List<Object>> stack;
+
+      private int depth;
+
+      private boolean initialized;
+
+      @Override
+      public void set(long integer) {
+
+         if (!initialized) {
+            output = new ArrayList<>();
+         }
+
+         output.add(integer);
+      }
+
+      @Override
+      public void set(double number) {
+
+         if (!initialized) {
+            output = new ArrayList<>();
+         }
+
+         output.add(number);
+      }
+
+      @Override
+      public void complete(int depth) {
+         if (depth > 0 && depth < this.depth) {
+            output = stack.pop();
+            this.depth--;
+         }
+      }
+
+      @Override
+      public void multi(int count) {
+
+         if (!initialized) {
+            output = new ArrayList<>(count);
+            initialized = true;
+         }
+
+         List<Object> a = new ArrayList<>(count);
+         output.add(a);
+         stack.push(output);
+         output = a;
+         this.depth++;
+      }
+   }
+
    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
       // Don't read any of the ByteBuf if we disabled reads
       if (disabledRead) {
          return;
       }
-      if (currentOutput == null) {
-         currentOutput = new PushOutput<>(ByteArrayCodec.INSTANCE);
-      }
       if (stateMachine.decode(in, currentOutput)) {
-         String type = currentOutput.getType().toUpperCase();
-         List content = currentOutput.getContent();
+         String type = currentOutput.type().toUpperCase();
          // Read a complete command, use a new one for next round
-         currentOutput = null;
-         List<byte[]> contentToUse = content.subList(1, content.size());
-         log.tracef("Received command: %s with arguments %s", type, Util.toStr(contentToUse));
+         List<byte[]> contentToUse = (List) currentOutput.getContent();
```

Comment:
Maybe silly, but can't this throw a cast exception? Or does the cast to `(List)` without types avoid that?

Reply:
It cannot as we are using a ByteArrayCodec, so the values will always be a byte[] underneath https://github.com/infinispan/infinispan/pull/10627/files#diff-598fcec6264f1bad944f54eb813e1f42a48cc1603daeee8d0a148cde18e15f4cR32
Unfortunately, lettuce doesn't have a way to pass the generics along. This code is basically the same as before, just do the cast on a different line.